crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -283,21 +200,29 @@ impl CheckpointExecutorEventLoop {
                                 "Bumping highest_executed_checkpoint watermark to {:?}",
                                 new_highest,
                             );
+
                             self.metrics.last_executed_checkpoint.set(new_highest as i64);
                             self.checkpoint_store
                                 .update_highest_executed_checkpoint(&checkpoint)
                                 .unwrap();
                         }
                         Some(committee) => {
-                            debug!(
-                                "Last checkpoint ({:?}) of epoch {:?} has finished execution",
-                                checkpoint.sequence_number(),
-                                checkpoint.epoch(),
+                            info!(
+                                ended_epoch = checkpoint.epoch(),
+                                last_checkpoint=checkpoint.sequence_number(),
+                                "Reached end of epoch",
                             );
                             self.checkpoint_store
                                 .update_highest_executed_checkpoint(&checkpoint)
                                 .unwrap();
-                            return Some((checkpoint, committee));
+
+                            // be extra careful to ensure we don't have orphans
+                            assert!(
+                                pending.is_empty(),
+                                "Pending checkpoint execution buffer should be empty after processing last checkpoint of epoch",
+                            );
+
+                            return self.create_committee_object(checkpoint, committee);
```

Comment:
Checkpoints can finish execution out of order, but items are removed from the pending queue in the same order that they are added (these are the semantics of `FutureOrdered`). And since we are adding checkpoints to the queue in order, they will come out ordered as well